omega/DecisionTree/DecisionTreeClassifier.py

```diff
@@ -0,0 +1,278 @@
+        gini = 0.0
+        for group in [left, right]:
+            y = group[1]
+            size = len(y) * 1.0
+            if size == 0:
+                continue
+            score = 0.0
+            for cl in classes:
+                proportion = sum(y == cl) / size
+                score += proportion ** 2
+            gini += (1.0 - score) * (size / num_instances)
+        return gini
+
+    def __test_split(self, index, value, x, y):
+        """Splits the dataset into two parts on the basis of a value.
+
+        Parameters
+        ----------
+        index :  int, index of the attribute on which to split
+        value : float, attribute value on which to split
+        x : ch.tensor, training data
+        y : ch.tensor, training labels
+        """
+        mask = ch.nonzero(ch.lt(x[:, index], value))
+        if mask.size():
+            mask = mask.view(mask.size()[0])
+            left = (ch.index_select(x, 0, mask), ch.index_select(y, 0, mask))
+        else:
+            left = (ch.DoubleTensor(), ch.DoubleTensor())
+
+        mask = ch.nonzero(1 - ch.lt(x[:, index], value))
+        if mask.size():
+            mask = mask.view(mask.size()[0])
+            right = (ch.index_select(x, 0, mask), ch.index_select(y, 0, mask))
+        else:
+            right = (ch.DoubleTensor(), ch.DoubleTensor())
+        return left, right
+
+    def fit(self, x, y):
+        """Trains the classifier on the training data and forms the decision tree
+
+        Parameters
+        ----------
+        x : ch.tensor
+            training data
+        y : ch.tensor
+            training labels
+
+        Returns
+        -------
+        root : dict
+            A dictionary containing the tree.
+        """
+        if isinstance(x, np.ndarray):
+            x = ch.from_numpy(x)
+        if isinstance(y, np.ndarray):
+            y = ch.from_numpy(y)
+        if self.__use_cuda:
+            x = x.cuda()
+            y = y.cuda()
+        self.root = self.__build_tree(x, y)
+        return self.root
+
+    def __build_tree(self, x, y):
+        """Initiates the tree building process by calling for the first split.
+
+        Parameters
+        ----------
+        x : ch.tensor
+            training data
+        y : ch.tensor
+            training labels
+
+        Returns
+        -------
+        root : dict
+            A dictionary containing the tree.
+        """
+        root = self.__get_split(x, y)
+        self.__split(root, 1)
+        return root
+
+    def __get_split(self, x, y):
+        """Finds the best split given the training data and labels
+
+        Parameters
+        ----------
+        x : ch.tensor
+            training data
+        y : ch.tensor
+            training labels
+
+        Returns
+        -------
+        dict : dict
+            A node for the tree with best split value given the data.
+        """
+        b_index, b_value, b_score, b_groups = 999, 999, 999, None
```

Comment:
What is this?

Reply:
random initializations, i just need the variables initialized outside the loop